test/async.coffee

```diff
@@ -0,0 +1,192 @@
+# Functions that contain the `await` keyword will compile into async
+# functions, which currently only Node 7+ in harmony mode can even
+# evaluate, much less run. Therefore we need to prevent runtimes
+# which will choke on such code from even loading it. This file is
+# only loaded by async-capable environments, so we redefine `test`
+# here even though it is based on `test` defined in `Cakefile`.
+# It replaces `test` for this file, and adds to the tracked
+# `passedTests` and `failures` arrays which are global objects.
+test = (description, fn) ->
+  try
+    fn.test = {description, currentFile}
+    await fn.call(fn)
+    ++passedTests
+  catch e
+    failures.push
+      filename: currentFile
+      error: e
+      description: description if description?
+      source: fn.toString() if fn.toString?
+
+
+# always fulfills
+winning = (val) -> Promise.resolve val
+
+# always is rejected
+failing = (val) -> Promise.reject new Error val
+
+
+test "async as argument", ->
+  ok ->
+    await winning()
+
+test "explicit async", ->
+  a = do ->
+    await return [1, 2, 3]
+  eq a.constructor, Promise
```

Comment:
Why are you comparing `a.constructor` and `Promise`? What exactly is this test testing, that `Promise`s are returned by `await`? You might want to make that clearer in the test description. At first glance I assume we should be testing that `a()` returns `[1, 2, 3]`.

What about another helper function like `winning` that returns some value, like `42`, that the tests can check against? So it’s not just that a promise is returned or that the promise resolves, but that the correct value is returned.